bibblio.py

```diff
@@ -0,0 +1,377 @@
+
+        representations = self._db.query(Representation)\
+            .join(Representation.resource).join(Resource.links)\
+            .join(Hyperlink.identifier).filter(
+                Identifier.id==identifier.id,
+                Hyperlink.rel==Hyperlink.OPEN_ACCESS_DOWNLOAD)\
+            .options(eagerload(Representation.resource))
+
+        text_representation = representations.filter(
+            Representation.media_type.in_(self.TEXT_MEDIA_TYPES),
+            Representation.content.isnot(None))\
+            .limit(1).all()
+
+        if text_representation:
+            # Get the full text if it's readily available.
+            [representation] = text_representation
+            full_text = self._html_to_text(representation.content)
+            full_text = self._shrink_text(full_text)
+            return full_text, representation.resource.data_source
+
+        # If it's gotta be an EPUB, make sure it matches the download url.
+        epub_representation = representations.filter(
+            Representation.media_type==Representation.EPUB_MEDIA_TYPE)\
+            .limit(1).all()
+
+        if not epub_representation:
+            # Access to the full text isn't available.
+            return None, None
+
+        [representation] = epub_representation
+        url = representation.url
+        content = representation.content
+        with EpubAccessor.open_epub(url, content=content) as (zip_file, package_path):
+            return (
+                self.extract_plaintext_from_epub(zip_file, package_path),
+                representation.resource.data_source
+            )
+
+    @classmethod
+    def extract_plaintext_from_epub(cls, zip_file, package_document_path):
+        spine, manifest = EpubAccessor.get_elements_from_package(
+            zip_file, package_document_path, ['spine', 'manifest']
+        )
+
+        text_basefiles = list()
+        for child in spine:
+            if child.tag == '{%s}itemref' % EpubAccessor.IDPF_NAMESPACE:
+                text_basefiles.append(child.get('idref'))
+
+        epub_item_elements = list()
+        for child in manifest:
+            if (child.tag == '{%s}item' % EpubAccessor.IDPF_NAMESPACE
+                and child.get('id') in text_basefiles):
+                epub_item_elements.append(child)
+
+        # Sort the items by their order in the spine.
+        epub_item_elements.sort(key=lambda el: text_basefiles.index(el.get('id')))
+
+        # Get the full EPUB filename for each text document.
+        text_filenames = [el.get('href') for el in epub_item_elements]
+        full_path = os.path.split(package_document_path)[0]
+        text_filenames = [os.path.join(full_path, f) for f in text_filenames]
+
+        accumulated_text = u''
+        for filename in text_filenames:
+            with zip_file.open(filename) as text_file:
+                raw_text = cls._html_to_text(text_file.read())
+                accumulated_text += raw_text
```

Comment:
I don't know if this matters but there should be a space  in between the text chunks so the point where two chunks meet isn't treated as a single word.